### PR TITLE
[Remote Clock] Wait for first tick and recalculate historical request bounds

### LINKED ISSUE #5221: Remote clock shows empty plots
#### Summary
There is a race condition with the remote clock whereby on initial load of the application plots will show no historical data due to the historical request being issued with start and end times of `0`. This is because the historical requests are issued before the remote clock has got a time back from the server.

There are two potential solutions here:
1. The remote clock should issue a proper bounds update (not a "tick") when it eventually gets a time from the server. 
2. The telemetry API blocks all historical requests if the remote clock has not yet received a valid time from the server.

2 Feels more correct to me because we can't really issue historical requests until we know the time. Issuing requests with bad start and end times is at best wasteful, and at worst may result in errors from backend systems.

#### Steps to Reproduce
1. Configure the remote clock
2. Select the remote clock in the time conductor
3. Navigate to a display with plots
4. Refresh the browser
5. Observe that the plots show no historical data. 
6. Switching to local clock and then back to the remote clock will fix the issue as the remote clock now has a valid time to use.

#### Environment
* Open MCT Version: 2.0.3
* Deployment Type: STM test environment (talk to @akhenry for details)

#### Impact Check List

- [x] Data loss or misrepresented data?
- [ ] Regression? Did this used to work or has it always been broken?
- [x] Is there a workaround available?
- [x] Does this impact a critical component?
- [ ] Is this just a visual bug with no functional impact?
- [ ] Does this block the execution of e2e tests?

### PR DESCRIPTION
**Pull request by ozyx**
Closes #5221 

### Describe your changes:

Adds a `TelemetryRequestInterceptor`. Block historical requests until the remote clock receives its first tick, and then use the interceptor to recalculate the request bounds.

Tests for `TelemetryRequestInterceptor` to be added in a future PR.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [x] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

File: src/api/telemetry/TelemetryAPI.js [hunk 1]
```diff
@@ -647,7 +589,104 @@ define([
         }
 
         return provider.getLimits(domainObject);
-    };
+    }
+}
+
+/**
```
Comment (akhenry) [inline src/api/telemetry/TelemetryAPI.js:595]: Nice one!

File: src/plugins/remoteClock/requestInterceptor.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * Open MCT includes source code licensed under additional open source
+ * licenses. See the Open Source Licenses file (LICENSES.md) included with
+ * this source code distribution or the Licensing information page available
+ * at runtime from the About dialog for additional information.
+ *****************************************************************************/
+
+function requestInterceptor(openmct, remoteClockIdentifier, waitForBounds) {
```
Comment (akhenry) [inline src/plugins/remoteClock/requestInterceptor.js:23]: Let's name what type of request interceptor it is. It's a `remoteClockRequestInterceptor`. Or something similar.

File: src/api/telemetry/TelemetryAPI.js [hunk 3]
```diff
@@ -276,7 +174,48 @@ define([
...
+     */
+    listRequestInterceptors(identifier, request) {
+        return this.requestInterceptorRegistry.getInterceptors(identifier, request);
+    }
+
+    /**
+     * Inovke interceptors if applicable for a given domain object.
+     * @private
+     */
+    async applyRequestInterceptors(identifier, request) {
```
[the remark this answers is not in the document]
Comment (ozyx) [reply src/api/telemetry/TelemetryAPI.js]: This is also used by `TelemetryCollection` so it shouldn't be marked as private

File: e2e/tests/plugins/telemetryTable/telemetryTable.e2e.spec.js [hunk 4]
```diff
@@ -71,25 +71,34 @@ test.describe('Telemetry Table', () => {
...
         await endTimeInput.fill('');
         await endTimeInput.fill(endDate);
         await page.keyboard.press('Enter');
 
         await expect(tableWrapper).not.toHaveClass(/is-paused/);
+
+        // Get the most recent telemetry date
+        const latestTelemetryDate = await page.locator('table.c-telemetry-table__body > tbody > tr').last().locator('td').nth(1).getAttribute('title');
+
+        // Verify that it is <= our new end bound
```
Comment (unlikelyzero) [inline e2e/tests/plugins/telemetryTable/telemetryTable.e2e.spec.js:99]: I would update the testcase name to include these additional checks

File: src/api/telemetry/TelemetryAPI.js [hunk 5]
```diff
@@ -330,14 +269,17 @@ define([
...
         if (!provider) {
             this.requestAbortControllers.delete(abortController);
 
             return this.handleMissingRequestProvider(domainObject);
         }
 
-        return provider.request.apply(provider, arguments)
+        let args = await this.applyRequestInterceptors(domainObject.identifier, arguments[1]);
+
+        return provider.request(domainObject, args)
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/api/telemetry/TelemetryAPI.js:281]: The issue here is that the _provider_ is no longer receiving all of the arguments passed into the `request` function. Previously it was.

**Review by akhenry — CHANGES_REQUESTED** [review]

**Review by unlikelyzero — APPROVED** [review]
e2e improvements look great!

If you have time before Andrew's review, you might want to put in a Remote Clock stub that captures this behavior as filed in the initial bug.

```js
// /e2e/tests/plugins/remoteClock
//boilerplate
test.describe('Remote Clock Tests', () => {
    test.fixme('Remote Clock accepts telemetry on initial load', async ({ page }) => {
    //addInitScript to with remote clock
    //Navigate directly to telemetry by URL OR by clicking time conductor
    };
}

**Review by akhenry — APPROVED** [review]

**Pull request merged**